examples/reshare/src/orchestrator/actor.rs

```diff
@@ -246,45 +267,111 @@ where
+                            );
+
+                            // Send the response back to the peer
+                            let response = wire::Message::Response(epoch, finalization);
+                            if orchestrator_sender
+                                .send(
+                                    Recipients::One(from),
+                                    response.encode().freeze(),
+                                    true,
+                                )
+                                .await
+                                .is_err()
+                            {
+                                warn!("failed to send orchestrator response, shutting down orchestrator");
+                                break;
+                            }
                         }
+                        wire::Message::Response(epoch, finalization) => {
+                            // Check if we actually requested this finalization from this peer
+                            match self.finalization_requests.entry(epoch) {
+                                Entry::Occupied(entry) if entry.get().0 == from => {
+                                    entry.remove();
+                                }
+                                _ => {
+                                    debug!(%epoch, ?from, "received unsolicited finalization, blocking peer");
+                                    self.oracle.block(from).await;
+                                    continue;
+                                }
+                            }
+
+                            // Look up the scheme to verify the certificate
+                            let Some(scheme) = self.scheme_provider.get_certificate_verifier(epoch) else {
+                                debug!(%epoch, ?from, "no scheme available to verify certificate");
+                                continue;
+                            };
+
+                            // Verify the certificate
+                            if !finalization.verify(
+                                &mut self.context,
+                                &scheme,
+                                &self.namespace,
+                            ) {
+                                debug!(
+                                    %epoch,
+                                    ?from,
+                                    "received finalization with invalid certificate, blocking peer"
+                                );
+                                self.oracle.block(from).await;
+                                continue;
+                            }
+
+                            debug!(
+                                %epoch,
+                                ?from,
+                                "verified requested finalization certificate, injecting finalization into marshal"
+                            );
+
+                            self.marshal.finalization(finalization).await;
```

Comment:
I suppose in the threshold case, we could really handle the latest finalization?